### PR TITLE
feat: implement runtime link name selection and a2a calls

### LINKED ISSUE #317: [RFC] Separate Actor and Provider SDKs
# [RFC] Separation of Guest and Provider SDKs

## Summary
This Request for Comment is soliciting feedback on the plan for separating the wasmCloud guest and provider SDKs. The suggested changes would create separate libraries for guests (actors) and capability providers and do so in a way that takes into account a component-based future while being an example of when not to create shared dependencies.

## Rationale
Today, we have [wasmbus_rpc](https://github.com/wasmCloud/weld/tree/main/rpc-rs), a library that provides an RPC foundation to allow capability providers to communicate with actors via hosts and perform routine functions like applying link definitions. Additionally, this exact same library also provides a foundation on which code is generated to give guest modules (actors) the ability to communicate bi-directionally with capability providers.

This creates an unfortunate tight coupling where both actors and providers share a common dependency or even a sub-tree of dependencies. The library is chock full of conditional compilation that checks for a `wasm` target architecture or non-wasm. At first this may not seem like that bad of a pattern because of the convenience of "one crate to rule them all", however, the real workflow-breaking problems occur with transitive dependencies. `wasmbus-rpc` forces both providers and actors to share a common dependency tree when the vast majority of projects don't need those common libraries.

### Problematic Use Cases
The most common problem case is when the provider and actor both depend on _different versions_ of the same shared dependency. This breaks builds in Rust and would also break builds for Go-based actors and providers.

One of the worst examples of how this pattern causes problems is the [lattice-controller](https://github.com/wasmCloud/capability-providers/tree/main/lattice-controller) capability provider, a provider that gives an actor the ability to interact with a lattice's management interface.

In this scenario, the lattice provider uses `wasmbus-rpc` which, in turn, uses a fixed version of `async-nats` (the NATS client library for Rust). However, the wasmCloud control interface is _also_ exposed over NATS, and so the capability provider needs to create its own instances of types in `async-nats`. At **best** this forces the provider and `wasmbus-rpc` to be in lock-step on the same version of the shared dependency, and at **worst** actually prevents features and bug fixes from being used by either of these crates. If **Go** follows the same shared common library pattern, it too will have conflicting NATS dependencies in the shared dependency tree.

There are myriad other scenarios that arise from having code generated from interface definitions that only works on specific versions of `wasmbus-rpc`, even though only a small portion of generated code needs access to RPC functionality, and even then, that is only needed by providers. Actor "RPC" with providers and other actors is really just a host call, which is managed entirely by the Wasm interface and not RPC.

## Implementation
This section describes how we intend to implement this separation of concerns and the impact to existing actors, providers, and generated code.

### wasmcloud-actor (Phase 1)
Today there are two so-called ABIs available to wasmCloud developers: the stable one, that we call the **wasmCloud ABI**, and the volatile one called **Wasm components** based on the WebAssembly component model. As the Wasm component standards and ecosystem stabilize, wasmCloud will move away from its own stable ABI toward embracing components by default.

The `wasmcloud-actor` Rust crate in this phase will provide developers with a consistent library used for building actor components _on the stable (legacy) ABI_. This will use smithy-generated code for data types that use the low-level guest primitives. 

⚠️ _Developers working with components will be left to their own devices for creating the appropriate wasifills, though we will support these efforts as best we can_.

For developer ergonomic benefits, this crate will come "out of the box" with easy to use wrappers around the most common interfaces:

* All wasmCloud builtins (number generation, logging, etc)
* Key-Value Store
* HTTP Server and Client
* Messaging
* Blob Store

Because the actors are using just the interfaces and not implementations, there's no waste overhead in including these by default. It also gives us the ability to optimize the experience of these for developer ergonomics in cases when the raw generated code is high friction.

❓**But isn't this more tight coupling** ❓ On the surface it does look like that. For example, we would have to issue a new version of the SDK if we change the messaging API even though the key value one didn't change. This looks like a violation of the single responsibility principle.

ℹ️ If we _absolutely need to patch just one of the core interfaces_ and releasing a new version of the wasmcloud SDK seems cumbersome (unlikely), then we can always have someone just generate from Smithy via `wasmcloud-bindgen` directly into their actor, effectively using an alternate core interface.

The flip side to that is if we keep our "core" API (limited to builtins + the most common contracts) at a slightly higher level than the smithy contracts, then we can actually _shield_ developers from volatility at the lower contract level.

Regardless, this tight coupling of the "core" modules will only be done during phase 1, so the question then becomes - is the added benefit of having more ergonomic interfaces on the core worth the short-term coupling we'll incur during phase 1? I think it is, because of the low risk of interface change during this phase.

_ℹ️ Rust will remove unused interfaces from compiled Wasm code during release builds._

### wasmcloud-provider (Phase 1)
During phase 1, this crate should largely represent a "copy and paste" effort from the existing `wasmbus-rpc` crate, just grabbing only the code that applies to capability providers. Note that today we have a `core` crate of shared data types that both actors and providers rely upon (e.g. the `LinkDefinition` and `Invocation` types). _This will go away_ and `core` types will be generated directly into this crate as needed.

The following is a list of responsibilities of this crate:

* Expose basic primitives for defining handlers and clients for contracts (the code "below" the codegen)
* Accept host configuration via `stdin` at provider start
* Communicate via NATS to send/receive invocations with actors
  * Includes signing and verifying invocation JWTs 
* Subscribe to changes in the appropriate key value store for link definitions
  * Ensure providers aren't invoked multiple times for the same link (de-dupe)

### wasmcloud-bindgen (Phase 1)
Today's wasmCloud developer environment relies on multiple crates within the [weld](https://github.com/wasmcloud/weld) repository. Not only is the name `weld` confusing to new developers, but recent creations like `smithy-bindgen` can be leveraged to help make the developer experience for third party interfaces be nearly as smooth as that of our baked-in first party interfaces. 

This phase of wasmcloud-bindgen will _only_ support Smithy-based interfaces and will be used to generate common data types, guest ABI wrappers, and provider handler wrappers. 

The choice as to which pieces are generated will be up to the developer. _Each individual project_ should use wasmcloud-bindgen inside `build.rs` with the appropriate parameters (e.g. actor vs provider) and _not_ create a shared crate. (remember the _share nothing_ principle).

Legacy code generation tasks from `build.rs` may be retired in favor of exclusively using the work now found in `smithy-bindgen`. 

_⚠️ Code generated with this crate will only work with the `wasmcloud-actor` or `wasmcloud-provider` crates, and will not be compatible with the legacy `wasmbus-rpc` crate_. The host should be unable to tell the difference, however.

### wasmcloud-actor (Phase 2)
This library will expose ergonomic wrappers around the core first party interfaces, with as few breaking changes from phase 1's API as possible. Underneath those interfaces, the ABI primitives will rely on the `actor` wit component/world interface for communicating with the host.

It will be easy for developers to use standard `wit` files, world files, and public Bytecode Alliance/Wasm standard tooling to generate the code necessary to work with wasmCloud.

### wasmcloud-provider (Phase 2)
This library supports all of the activities required by the provider crate in phase 1. This library will continue to support all RPC activities between hosts, providers, and actors. If there are pieces of code that need to be upgraded to use `wit`, then that will be done here. If changes are handled properly by the host(s) and components, this crate should have relatively few breaking changes during the transition to phase 2 since the provider<->host RPC doesn't use any Wasm primitives.

### wasmcloud-bindgen (Phase 2)
Unless there are specific convenience things that we still want in this crate, we can remove most (if not all) of the code from it. Instead, `wash` will be upgraded such that `wash build` will know how to do all of the component linking and management to build the right kind of components that will be compatible with this phase and the host(s) released during this phase.

During this phase if there are [wit-bindgen](https://github.com/bytecodealliance/wit-bindgen) language bindings that either need work or don't exist, then we will evaluate whether it makes sense for us to create or contribute those.

If this crate remains, it cannot be _required_ to build code for wasmCloud during this phase. It can only make things _easier_ or _simpler_.

**Requirement** - using nothing but Bytecode Alliance/standard Wasm tooling, developers must be able to create an actor that can work in a wasmCloud host. Providers will still be Rust-first and will continue to rely on the existing `wasmcloud-provider` (phase 2) crate.

## Timeline and Gate Conditions
The following outlines the planned road from today to the ideal end state for this effort.

### Phase 1
All actors and providers using these crates will be compatible with Elixir OTP Host v0.62+ without modification. The ABI expected by the host from actors and the RPC mechanism for providers will not change.

During this phase, early adopters looking to work with components on their own will be responsible for managing their own **wasifills** and linking using BA tools. wasmCloud will make no support guarantees for components, though we will of course attempt to remain as far up on the leading edge of component development as possible with **wasmtime**.

* **wasmcloud-actor (p1)** - isolated crate providing ergonomic wrappers and core primitives for actors
* **wasmcloud-provider (p1)** - isolated crate providing ergonomic wrappers and core primitives for providers
* **wasmcloud-bindgen (p1)** - crate that can be used to generate code from **Smithy** contracts suitable for use by actors or providers that depends on primitives from the preceding 2 crates

Additional changes that will take place during phase 1 include the removal of the use of `async` code from within an actor for synchronous activities. We will also probably no longer have actors respond to a health request. There are a number of reasons for this, but the short version is that stateless actors have no health: they're either available or not.

Here is an example of what an actor might look like on phase 1 libraries to build our standard key-value counter:

```rust
use wasmcloud_actor::{kv, http};

#[handles(HttpServer)]
struct MyDemo;

impl HttpServer for MyDemo {
    fn handle_request(request: &HttpRequest) -> Result<HttpResponse> {
        let trimmed_path: Vec<&str> = req.path.trim_matches('/').split('/').collect();
        
        let count = match (req.method.as_ref(), trimmed_path.as_slice()) {
            ("GET", ["api", "counter", counter]) => {
                kv::increment(counter, 1)?
            },
            _ => 0
        };
        
        let output = json!({
            "counter": count
        });
        Ok(HttpResponse::ok(output))
    }    
}
```

Note how the boilerplate and ceremony is gone from the use of `kv`. If an actor needs to communicate with a different [link name](https://wasmcloud.com/docs/reference/host-runtime/links), then that can be set by calling a new function:

```rust
set_link_name("wasmcloud:keyvalue", "cache");

// do work on `cache` key value store
kv::increment("alice", 5)?;

reset_links(); // reset to "default"
```

The use of this new function to set link names will prepare developers for when this kind of thing becomes mandatory when using the component model (since `wit` doesn't know what a link name is).

Here's what it might look like to add in the use of a third-party contract generated from Smithy via `wasmcloud-bindgen (p1)`. In this sample, we'll use the `wasmcloud:sensors` contract to report the current temperature of the BBQ at request time:

```rust
use wasmcloud_actor::http; // stock interface
use sensors::*; // generated module

mod sensors; // bindgen-generated into src/sensors.rs

#[handles(HttpServer)]
struct BbqDemo;

impl HttpServer for BbqDemo {
    fn handle_request(request: &HttpRequest) -> Result<HttpResponse> {
        
        let value = query_sensor(SensorQuery {
            network_id: Some("home".to_string()),
            sensor_id: "BBQ_ONE".to_string()
        })?.data[0]; // e.g. 375.31 (assuming fahrenheit)
        
        let output = json!({
            "temperature": value
        });
        
        Ok(HttpResponse::ok(output))
    }    
}
```

Note that the sensors code is generated _directly_ into this actor, and is not part of a shared library.

#### Actor-to-Actor Calls
Actors calling actors will still be a very common, fully supported scenario. Our own [pet clinic](https://github.com/wasmCloud/examples/tree/main/petclinic) sample uses the API gateway pattern with an actor calling other actors depending on their role.

Today's code uses a rather clunky way of telling the host that the target is an actor rather than a capability provider. While preparing for `wit` based contracts (e.g. non-wasmCloud-proprietary), we can set the stage using the `set_target` function:

```rust
set_target("petclinic/visits");
visits::record_visit(Visit {
    owner: "kevin".to_string(),
    pet: "sage".to_string(),
    . . .
})?;
clear_target();
```

This function sets the current actor target for all subsequent host calls. That target will remain the same until either set to something else or cleared via `clear_target`. Setting targets this way will help pave the way for when we have "off the shelf" wit contract support.

### Phase 2
This phase will use the component model ABI and `wit`. The crates in this phase _are not_ backwards compatible with the Smithy-based legacy wasmCloud ABI. 

If developers want to build Smithy-based projects, they will need to continue to use deprecated Phase 1 tooling. 

The Elixir OTP host v0.63+ will continue to support the wasmCloud legacy API (deprecated) and the new component ABI by virtue of the Rust `wasmcloud` crate. As of phase 2, there will be no wasmCloud-proprietary code. The smithy "walled garden" will be gone.

* **wasmcloud-actor (p2)** - isolated crate providing ergonomic wrappers and wit/component-based core primitives for actors
* **wasmcloud-provider (p2)** - isolated crate providing ergonomic wrappers and core primitives for providers
* **wasmcloud-bindgen (p2)** - crate that may largely go away and be replaced by publicly available `wit` tooling and `wash build`

The following is an example of what code might look like for Rust guests on phase 2 (which should be indistinguishable from "regular" component code).

```rust
wit_bindgen::generate!("httpserver.wit");

use httpserver::*;

struct MyServer;

impl HttpServer for MyServer {
    fn handle_request(request: &HttpRequest) -> Result<HttpResponse, String> {
        Ok(HttpResponse::default())
    }
}

export_host!(MyServer);
```

### Phase 3
Elixir OTP host (and any other host) no longer supports smithy-based contracts and associated code generation. Deprecated entities from phase 2 will no longer be supported. In other words, as of phase 3, the Rust `wasmcloud` crate will only support components and will no longer recognize the legacy wasmCloud ABI. It is for this reason that there could conceivably be a pretty long gap between Phase 2 and Phase 3 while we await component stabilization in W3C specs and Bytecode Alliance tooling.

## Other Languages, C, and FFI
We were originally thinking of exposing the various SDKs as a C-compliant FFI layer so that other languages could take advantage of the shared code. However, since the ultimate goal is to "simply" use components and `wit`, the effort that would otherwise be placed here will likely be spent making sure that the `wit-bindgen` experience is smooth and supports all of the component-capable Wasm-targeting languages.

Creating a common C FFI layer that allows Go libraries to utilize the non-Wasm RPC functionality from Rust is out of scope for this RFC, though that work may be done during Phase 2 or 3.

### LINKED ISSUE #463: Multiple links per contract ID
Currently, the host only stores a single link per contract ID, it should be able to store multiple and take link name into account.
First, we should figure out how this would work in component model, refs https://github.com/wasmCloud/wasmCloud/issues/461#issuecomment-1673678571

### LINKED ISSUE #494: [Rust Host] Actor (module) -> provider invocations don't make it to the bus
From local debugging/print lines, the await on [read_to_end](https://github.com/wasmCloud/wasmCloud/blob/main/crates/host/src/wasmbus/mod.rs#L426) never resolves, so invocations don't make it onto `wasmbus.rpc`

Example to reproduce:
```
wash start actor wasmcloud.azurecr.io/kvcounter:0.4.2
wash start provider wasmcloud.azurecr.io/httpserver:0.19.1
wash start provider wasmcloud.azurecr.io/kvredis:0.22.0

wash link put MCFMFDWFHGKELOXPCNCDXKK5OFLHBVEWRAOXR5JSQUD2TOFRE3DFPM7E VAG3QITQQ2ODAOWB5TTQSDJ53XK3SHBEIFNK4AYJ5RKAX2UNSCAPHA5M wasmcloud:httpserver address=0.0.0.0:1234

wash link put -x default -p 4222 MCFMFDWFHGKELOXPCNCDXKK5OFLHBVEWRAOXR5JSQUD2TOFRE3DFPM7E VAZVC4RX54J2NVCMCW7BPCAHGGG5XZXDBXFUMDUXGESTMQEJLC3YVZWB wasmcloud:keyvalue

curl 127.0.0.1:1234/api/counter
```

If you `nats sub wasmbus.rpc.>`, you'll see the actor module invocation on the bus, but won't see the provider call

### PR DESCRIPTION
**Pull request by rvolosatovs**
## Feature or Problem

Define the API for https://github.com/wasmCloud/wasmCloud/issues/469 and implement

## Related Issues

Closes https://github.com/wasmCloud/wasmCloud/issues/469
Closes #463 
Closes #317 
Closes #494 (please test)

## Testing

Built on platform(s)
- [x] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [ ] aarch64-darwin
- [ ] x86_64-windows

Tested on platform(s)
- [x] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [ ] aarch64-darwin
- [ ] x86_64-windows

### CODE REVIEW

**Comment by thomastaylor312** [comment]
@rvolosatovs So I think this is close but there might have been some confusion when we last synced on this offline. I think it will be much cleaner if we keep the current `call` stuff exactly as it is, but expose a new interface that can included in the actor world.

```
// This is just a guess at names. Not strongly opinionated on interface or package name
package wasmcloud:lattice

interface lattice {
    variant actor-identifier {
        /// Actor public key
        public-key(string),
        /// Actor call alias
        alias(string),
    }

    // Note that this doesn't impose any restraints on the interface/contract name as we need to be able to support custom contracts
    set-link-name: func(interface: string, link-name: string)

    set-target: func(actor: actor-identifier)

    clear-link-name: func()

    clear-target: func()
}
```

Basically the host will provide that functionality and then we don't have to worry about passing interfaces or link names in the host call. What I envision the code looking like from an actor perspective is something like this:

```rust
set_link_name("wasmcloud:mycustom", "foobar");
// This will now be using the link name under the hood
custom::thing_i_call("blah");
clear_link_name();
```

An actor to actor call would be similar:

```rust
set_target(ActorIdentifier::CallAlias("foobar");
actor::thing_i_call("blah");
clear_target();
```

In the host implementation it would be something like this (very psuedocode-y, just trying to explain reasoning):

```rust
impl Lattice for Host {
    fn set_link_name(interface: String, link_name: String) {
        let current_interface = self.interfaces.get(interface);
        current_interface.link_name = link_name;
        self.interfaces.insert(interface, current_interface);
    }
}
    

/// Then when host.call comes
let current = self.interfaces.get(interface);
let inv = Invocation {
    link_name: current.link_name.unwrap_or("default")
    // All the rest of the stuff
}
```

Does that make any sense?

**Comment by rvolosatovs** [comment]
> @rvolosatovs So I think this is close but there might have been some confusion when we last synced on this offline. I think it will be much cleaner if we keep the current `call` stuff exactly as it is, but expose a new interface that can included in the actor world.
> 
> ```
> // This is just a guess at names. Not strongly opinionated on interface or package name
> package wasmcloud:lattice
> 
> interface lattice {
>     variant actor-identifier {
>         /// Actor public key
>         public-key(string),
>         /// Actor call alias
>         alias(string),
>     }
> 
>     // Note that this doesn't impose any restraints on the interface/contract name as we need to be able to support custom contracts
>     set-link-name: func(interface: string, link-name: string)
> 
>     set-target: func(actor: actor-identifier)
> 
>     clear-link-name: func()
> 
>     clear-target: func()
> }
> ```
> 
> Basically the host will provide that functionality and then we don't have to worry about passing interfaces or link names in the host call. What I envision the code looking like from an actor perspective is something like this:
> 
> ```rust
> set_link_name("wasmcloud:mycustom", "foobar");
> // This will now be using the link name under the hood
> custom::thing_i_call("blah");
> clear_link_name();
> ```
> 
> An actor to actor call would be similar:
> 
> ```rust
> set_target(ActorIdentifier::CallAlias("foobar");
> actor::thing_i_call("blah");
> clear_target();
> ```
> 
> In the host implementation it would be something like this (very psuedocode-y, just trying to explain reasoning):
> 
> ```rust
> impl Lattice for Host {
>     fn set_link_name(interface: String, link_name: String) {
>         let current_interface = self.interfaces.get(interface);
>         current_interface.link_name = link_name;
>         self.interfaces.insert(interface, current_interface);
>     }
> }
>     
> 
> /// Then when host.call comes
> let current = self.interfaces.get(interface);
> let inv = Invocation {
>     link_name: current.link_name.unwrap_or("default")
>     // All the rest of the stuff
> }
> ```
> 
> Does that make any sense?

I disagree with your proposal, let's discuss it in detail offline tomorrow.

File: crates/actor/src/lib.rs
Comment (brooksmtownsend) [inline crates/actor/src/lib.rs:77]: Flagging because I see `unwrap` but I'm fine with this if it's just temporary code.

File: crates/host/src/wasmbus/mod.rs
Comment (brooksmtownsend) [inline crates/host/src/wasmbus/mod.rs:348]: nit: would be good to mark this and other TODOs as `TODO(#469)`, some GitHub plugins can link that up with the PR and later we can search for all the relevant TODOs

**Review by brooksmtownsend — DISMISSED** [review]

File: crates/actor/src/lib.rs
Comment (rvolosatovs) [reply crates/actor/src/lib.rs:77]: That's a test and this is never actually called

File: crates/runtime/wit/deps/cli/exit.wit
Comment (thomastaylor312) [inline crates/runtime/wit/deps/cli/exit.wit:2]: Nit: Looks like a fat fingered key

**Review by thomastaylor312 — COMMENTED** [review]
This is good to go from a code review perspective. Not giving the full approval so @connorsmith256 can give final LGTM once he tests that #494 is resolved

**Comment by connorsmith256** [comment]
I haven't reviewed the updated code yet, but it looks like #494 is still happening :( I followed the steps to reproduce in the issue, and still don't see an outbound RPC call from the actor to the provider

File: tests/wasmbus.rs
Comment (connorsmith256) [inline tests/wasmbus.rs:393]: Why switch to magic strings?

File: crates/host/src/wasmbus/mod.rs
Comment (connorsmith256) [inline crates/host/src/wasmbus/mod.rs:283]: Consider a named constant, to disambiguate among references to default link names and default lattice prefixes

File: crates/host/src/wasmbus/mod.rs
Comment (connorsmith256) [inline crates/host/src/wasmbus/mod.rs:2522]: Nice, thanks for fixing this

**Review by connorsmith256 — CHANGES_REQUESTED** [review]
No big issues, only requesting changes until outbound invocations from modules are confirmed to work. I suggest we add to the wasmbus tests: start the kvcounter actor, increment via HTTP request, and confirm the data in redis is updated

File: crates/host/src/wasmbus/mod.rs
Comment (rvolosatovs) [reply crates/host/src/wasmbus/mod.rs:283]: I don't really see what's there to disambiguate, `links` is keyed by link names, not lattice prefixes

File: tests/wasmbus.rs
Comment (rvolosatovs) [reply tests/wasmbus.rs:393]: For clarity, simplifying debugging and being explicit.
Note, that in doing so I've fixed an invalid link name specified in the link deletion.

File: crates/host/src/wasmbus/mod.rs
Comment (connorsmith256) [reply crates/host/src/wasmbus/mod.rs:283]: What I mean is distinguishing between semantics of a common string (`"default"`) in the repo. Right now, if someone wants to find all the places we fall back on `default` for a link name, they have to search for this string, and then go through the mental effort of excluding uses where `default` was being used in the context of a lattice prefix. (I had to go through this effort a week ago when adding support for custom lattice prefixes and custom link names)

Swapping to named constants like `DEFAULT_LINK_NAME` and `DEFAULT_LATTICE_PREFIX` would remove any ambiguity

File: tests/wasmbus.rs
Comment (connorsmith256) [reply tests/wasmbus.rs:393]: Not going to block on this since it's a test, but my $0.02: copy-paste errors are copy-paste errors, regardless of whether they're a named var or a string. The advantage of a named var over a string is to prevent fat-finger typos, like passing `htpserver` as a link name when it should be `httpserver`. If it's a named variable, typos won't compile

Comment (rvolosatovs) [reply tests/wasmbus.rs:393]: Sure, the advantage of plain strings is maintenance, this string has to be duplicated in at least 5 different crates and indirection here just makes it more difficult to keep track of these. Looking at a single call in isolation should be enough to figure out what link name is used

Comment (connorsmith256) [reply tests/wasmbus.rs:393]: Did you mean to respond to the other thread? I'm not sure I follow the reference to multiple crates, since this is all in one test file

File: crates/runtime/wit/deps/cli/exit.wit
Comment (rvolosatovs) [reply crates/runtime/wit/deps/cli/exit.wit:2]: Indeed, but we should not fix a dependency

**Comment by rvolosatovs** [comment]
@connorsmith256 I cannot reproduce #494 on this branch:
```
$ wash start actor wasmcloud.azurecr.io/kvcounter:0.4.2

Actor [MCFMFDWFHGKELOXPCNCDXKK5OFLHBVEWRAOXR5JSQUD2TOFRE3DFPM7E] (ref: [wasmcloud.azurecr.io/kvcounter:0.4.2]) started on host [NDCCV4YKFKHUMTRFVQ6ZU4TQRP7ZKO236TF5XICAQVB32RHHN6FMAQ6D]
$ wash start provider wasmcloud.azurecr.io/httpserver:0.19.1

Provider [VAG3QITQQ2ODAOWB5TTQSDJ53XK3SHBEIFNK4AYJ5RKAX2UNSCAPHA5M] (ref: [wasmcloud.azurecr.io/httpserver:0.19.1]) started on host [NDCCV4YKFKHUMTRFVQ6ZU4TQRP7ZKO236TF5XICAQVB32RHHN6FMAQ6D]
$ wash start provider wasmcloud.azurecr.io/kvredis:0.22.0

Provider [VAZVC4RX54J2NVCMCW7BPCAHGGG5XZXDBXFUMDUXGESTMQEJLC3YVZWB] (ref: [wasmcloud.azurecr.io/kvredis:0.22.0]) started on host [NDCCV4YKFKHUMTRFVQ6ZU4TQRP7ZKO236TF5XICAQVB32RHHN6FMAQ6D]
$ wash link put MCFMFDWFHGKELOXPCNCDXKK5OFLHBVEWRAOXR5JSQUD2TOFRE3DFPM7E VAG3QITQQ2ODAOWB5TTQSDJ53XK3SHBEIFNK4AYJ5RKAX2UNSCAPHA5M wasmcloud:httpserver address=0.0.0.0:1234

Published link (MCFMFDWFHGKELOXPCNCDXKK5OFLHBVEWRAOXR5JSQUD2TOFRE3DFPM7E) <-> (VAG3QITQQ2ODAOWB5TTQSDJ53XK3SHBEIFNK4AYJ5RKAX2UNSCAPHA5M) successfully
$ wash link put -x default -p 4222 MCFMFDWFHGKELOXPCNCDXKK5OFLHBVEWRAOXR5JSQUD2TOFRE3DFPM7E VAZVC4RX54J2NVCMCW7BPCAHGGG5XZXDBXFUMDUXGESTMQEJLC3YVZWB wasmcloud:keyvalue

Published link (MCFMFDWFHGKELOXPCNCDXKK5OFLHBVEWRAOXR5JSQUD2TOFRE3DFPM7E) <-> (VAZVC4RX54J2NVCMCW7BPCAHGGG5XZXDBXFUMDUXGESTMQEJLC3YVZWB) successfully
$ curl 127.0.0.1:1234/api/counter
{"counter":1}
$ curl 127.0.0.1:1234/api/counter
{"counter":2}
$ curl 127.0.0.1:1234/api/counter
$ redis-cli
127.0.0.1:6379> keys *
1) "counter:default"
```

File: tests/wasmbus.rs
Comment (rvolosatovs) [reply tests/wasmbus.rs:418]: See the updated diff, names must match in 3 Wasm test binaries and in 2 tests - one you commented upon and a `wasmcloud_runtime`-level test

**Review by connorsmith256 — APPROVED** [review]
Love all the new test cases ❤️

**Pull request merged**
